crates/ruff_python_parser/src/string_token_flags.rs

```diff
@@ -290,65 +344,72 @@ impl fmt::Debug for StringKind {
 
 impl From<StringKind> for ruff_python_ast::StringLiteralFlags {
     fn from(value: StringKind) -> ruff_python_ast::StringLiteralFlags {
-        debug_assert!(!value.is_f_string());
-        debug_assert!(!value.is_byte_string());
-
         let mut new = ruff_python_ast::StringLiteralFlags::default();
         if value.quote_style().is_double() {
             new = new.with_double_quotes();
         }
         if value.is_triple_quoted() {
             new = new.with_triple_quotes();
         }
-        new.with_prefix({
-            if value.is_u_string() {
-                debug_assert!(!value.is_raw_string());
-                StringLiteralPrefix::UString
-            } else if value.is_raw_string() {
-                StringLiteralPrefix::RString
-            } else {
-                StringLiteralPrefix::None
+        new.with_prefix(match value.prefix() {
+            None => StringLiteralPrefix::None,
+            Some(StringPrefix::Unicode) => StringLiteralPrefix::UString,
+            Some(StringPrefix::Raw { uppercase: true }) => {
+                StringLiteralPrefix::RString { uppercase: true }
+            }
+            Some(StringPrefix::Raw { uppercase: false }) => {
+                StringLiteralPrefix::RString { uppercase: false }
+            }
+            Some(StringPrefix::Bytes | StringPrefix::RawBytes { .. }) => {
+                panic!("Attempting to convert a bytestring into a non-bytestring!")
+            }
+            Some(StringPrefix::Format | StringPrefix::RawFormat { .. }) => {
+                panic!("Attempting to convert an f-string into a non-fstring!")
```

Comment:
It's unfortunate that we need to use `panic` here but I understand the reasoning behind it which is that we're going from a wide type containing all possible string prefix to a narrow one for a specific type of string (for example `StringPrefix` to `FStringPrefix`).

I think this should actually be an `unreachable` instead as our lexer will raise an error for invalid combination of prefixes used.

I'm sure you must have thought of this but if possible, I would try to find any type-safe way to do this conversion to avoid a potential panic.

Reply:
I agree this is somewhat unfortunate. It might be worth noting, though, that this isn't really any less type-safe than previously -- the `panic!`s in this function are really just doing the work for us that the `debug_assert!`s were previously.

The only other way I can _see_ of doing this would be to implement `TryFrom` instead of `From` here -- return an `Err` variant instead of panicking. But that feels incorrect to me. If we hit the panicking branch, I think we _should_ be immediately bailing from this function, since that indicates a logical error in our code somewhere. If we've written the tokenizer and parser correctly, it should be impossible for us to get here.

Reply:
> the `panic!`s in this function are really just doing the work for us that the `debug_assert!`s were previously.

You're correct although they differ from a user perspective. `debug_assert!` will only panic on debug builds while `panic` will bail out even on the release version.

> I think we _should_ be immediately bailing from this function, since that indicates a logical error in our code somewhere. If we've written the tokenizer and parser correctly, it should be impossible for us to get here.

Yes, this is the reason I suggest using the `unreachable` macro instead. As a developer it's hard to reason about a `panic` usage unless there's an indicator like a comment. However, if `unreachable` is used, one can say that this is indeed a logical error.

Reply:
Makes sense — I'll switch to `unreachable`, thanks

Reply:
> Makes sense — I'll switch to `unreachable`, thanks

Done in https://github.com/astral-sh/ruff/pull/10314/commits/3ee6454f02aa49f2ee451da19f704e280c248dc0

Reply:
One way in which we could make this type-safe would be to emit different tokens for string literals and bytes literals. Then we would be able to have the same three-way `f-string/bytestring/everything-else` distinction at the token level as we do at the AST level, which would mean we would be able to have three different bitflags at the token level. We wouldn't need to do the type narrowing here because the type would already be narrowed.

But that would be quite a large change to make.